In [1]:
import sys
sys.path.append('/home/wjb31/src/openfst//openfst-1.6.3/INSTALL_DIR/lib/python2.7/site-packages/')
sys.path.append('/home/wjb31/src/openfst//specializer-master/')
sys.path.append('/home/wjb31/MLSALT/util/python/')
import pywrapfst as fst
import specializer
import math
import string
import utilfst

In [2]:
DIR='/home/wjb31/MLSALT/MLSALT3/practical/recasing/'

PTB_DEVWORDS='/home/wjb31/MLSALT/MLSALT3/practical/recasing/data/ptb/ptb-dev.words'
PTB_DEVCHARS='/home/wjb31/MLSALT/MLSALT3/practical/recasing/data/ptb/ptb-dev.chars'
PTB_DEVINPUT='/home/wjb31/MLSALT/MLSALT3/practical/recasing/data/ptb/ptb-dev.input.fsts/'

KFTT_DEVWORDS='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-dev.words'
KFTT_DEVCHARS='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-dev.chars'
KFTT_DEVINPUT='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-dev.input.fsts/'

KFTT_TESTWORDS='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-test.words'
KFTT_TESTCHARS='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-test.chars'
KFTT_TESTINPUT='/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/kftt/kftt-test.input.fsts/'

chars_ST=fst.SymbolTable.read_text('/home/wjb31/MLSALT/MLSALT3/practical/recasing/data/chars.syms')
words_ST=fst.SymbolTable.read_text('/home/wjb31/MLSALT/MLSALT3/practical/recasing/data/words.syms')

## Questions
__Question 1.__
The composition of `PTB_Dev21` and `toUpper` yields a transducer. Describe the input language and the output language of this transducer.

__Question 2.__
The composition of `PTB_Dev21` and `toUpper_lm_char_2` also yields a transducer. How does its input language and its output language differ from those of Question 1?

In [3]:
PTB_Dev21 = fst.Fst.read(PTB_DEVINPUT+'21.fst')
#utilfst.printstrings(PTB_Dev21, syms=chars_ST)
# first, make an fst compiler for characters
char_compiler = fst.Compiler(isymbols=chars_ST, osymbols=chars_ST, keep_isymbols=True, keep_osymbols=True)
# Now build this transducer from the arc-format text file.   
# N.B. build this transducer from this text file in this way to avoid any variations due to local text environments
with open('/home/wjb31/MLSALT/MLSALT3/practical/recasing//common/toUpper.txt') as fp:
    for line in fp:
        print >> char_compiler, line

toUpper=char_compiler.compile()

In [4]:
utilfst.printstrings(fst.compose(PTB_Dev21, toUpper), syms=chars_ST, nshortest=5) # input language


["<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>"]

In [5]:
utilfst.printstrings(fst.compose(PTB_Dev21, toUpper), syms=chars_ST, nshortest=5, project_output=True) # output language

["<s> W E ' R E _ A B O U T _ T O _ S E E _ I F _ A D V E R T I S I N G _ W o R K s . </s>",
 "<s> W E ' r E _ A B O U T _ T O _ S E E _ I F _ A D V E R T I s I N G _ W o R K s . </s>",
 "<s> W E ' R E _ A B O U T _ T O _ S E E _ I F _ A D v E R T I S I N G _ W o R K s . </s>",
 "<s> W E ' R E _ A B O U T _ T O _ S E E _ I F _ A D V E R T I S I N G _ W O R k S . </s>",
 "<s> W e ' r E _ A B O U T _ T O _ S E E _ I F _ A D V E R T I s I N G _ W o R K s . </s>"]

#### Answer 1:
* the input language is lowercase character strings
* the output language is mixed-cased character strings

In [6]:
toUpper_lm_char_2 = fst.Fst.read('toUpper.lm.char.2.fst')

In [7]:
utilfst.printstrings(fst.compose(PTB_Dev21, toUpper_lm_char_2), syms=chars_ST, nshortest=5) # input language

["<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> w e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>"]

In [8]:
utilfst.printstrings(fst.compose(PTB_Dev21,toUpper_lm_char_2),syms=chars_ST,nshortest=5,project_output=True)#output language

["<s> W e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> W e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> W e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>",
 "<s> W e ' r e _ a b o u t _ t o _ s e e _ i f _ a d v e r t i s i n g _ w o r k s . </s>"]

#### Answer 2:
* The input language is all lowercase character string
* The output language is a character string with initial capital character in a sentence

## Questions

__Question 3.__ Section 4.5 gives the following sequence of operations for applying the word-based unigram language model

   `PTB_Dev21_mixed = fst.compose(PTB_Dev21, toUpper).project(project_output=True)`

   `PTB_Dev21_cseq2wseq = fst.compose(PTB_Dev21_mixed, cseq2wseq).rmepsilon().arcsort()`

Explain why the output projection operation is needed.

__Question 4.__ 
Section 4.3.1 states that

   `We build a single character to word transducer that accepts words both in-vocabulary and out-of-vocabulary words; the former are mapped to their cased versions as found in the training data, and the latter are mapped to <unk>:`

a) Explain whether or not this is actually done by the suggested union operation of the `c2w` and `unk` transducers. 

b) Using failure transitions, draw (by hand) a modified version of the transducer of this machine that preserves its transductions and also maps every other sequence from `{a-Z}+` to `<unk>`.

#### Answer 3:
* the word sequence is generated from mixed character sequence and we only need the output of the transducer.

#### Answer 4:
* (a) No, it's not achieved by union. Because the if a unknown word shows up in the sequence the transducer  


## Questions

__Question 5.__
Using the word-level and the character-level lattices, perform recasing under the combined word-level and character-level n-gram language model scores.

a) Complete the table in the Results section.

b) Give the sequence of operations that you use for the different model combination schemes.

c) Compare performance at the character level and word level of the word and character n-gram language models.

### Character-based Bi-gram model - KFTT-test

In [12]:
rawhyps = []
for id in xrange(1,1113): # KFTT-test has 1112 sentences
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_2)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:

with open(KFTT_TESTCHARS) as f:
    kfttcrefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL Total accuracy: 94.805737 (122617/129335)
94.8057370395
None
WORD-LEVEL Total accuracy: 75.483156 (15662/20749)
75.4831558147
None


### Word-based Uni-gram model - KFTT-dev

In [9]:
cseq2wseq = fst.Fst.read('cseq2wseq.fst')
lm_word_1 = fst.Fst.read('lm.word.1.fst')

In [19]:
rawhyps = []
for id in xrange(1,1114):
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst') # read input FSA with lower case character sequences
    x1 = fst.compose(ifst, toUpper).project(project_output=True) #project to mixed case character sequences and project onto the output
    x2 = fst.compose(x1, cseq2wseq) # compose with the chara83.01cter sequence to word sequence transducer
    ofst = fst.compose(x2, lm_word_1) # compose with the unigram language model
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST)) # get the best mixed case character sequence under the unigram language model 

chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:
with open(KFTT_DEVCHARS) as f:
    kfttcrefs = [line.strip() for line in f]    
with open(KFTT_DEVWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL Total accuracy: 83.008507 (103241/124374)
83.0085066011
None
WORD-LEVEL Total accuracy: 73.029234 (14489/19840)
73.029233871
None


### Word-based Uni-gram model - KFTT-test

In [20]:
rawhyps = []
for id in xrange(1,1113):
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst') # read input FSA with lower case character sequences
    x1 = fst.compose(ifst, toUpper).project(project_output=True) #project to mixed case character sequences and project onto the output
    x2 = fst.compose(x1, cseq2wseq) # compose with the character sequence to word sequence transducer
    ofst = fst.compose(x2, lm_word_1) # compose with the unigram language model
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST)) # get the best mixed case character sequence under the unigram language model 

chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:
with open(KFTT_TESTCHARS) as f:
    kfttcrefs = [line.strip() for line in f]    
with open(KFTT_TESTWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL Total accuracy: 81.775235 (105764/129335)
81.7752348552
None
WORD-LEVEL Total accuracy: 70.417851 (14611/20749)
70.4178514627
None


### Character-based Bigram + Word-based Uni-gram model - PTB-dev

In [24]:
rawhyps = []
for id in xrange(1,1701): 
    ifst = fst.Fst.read(PTB_DEVINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper_lm_char_2).project(project_output=True)
    x2 = fst.compose(x1, cseq2wseq)
    ofst = fst.compose(x2, lm_word_1)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST))
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the PTB-dev character and word references:
with open(PTB_DEVCHARS) as f:
    ptbcrefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    ptbwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(ptbcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(ptbwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 98.791305 (206378/208903)
98.7913050555
None
WORD-LEVEL:  Total accuracy: 94.555991 (32306/34166)
94.5559913364
None


### Character-based Bigram + Word-based Uni-gram model - KFTT-dev

In [22]:
rawhyps = []
for id in xrange(1,1114): 
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper_lm_char_2).project(project_output=True)
    x2 = fst.compose(x1, cseq2wseq)
    ofst = fst.compose(x2, lm_word_1)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST))
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:

with open(KFTT_DEVCHARS) as f:
    kfttcrefs = [line.strip() for line in f]
with open(KFTT_DEVWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 94.495634 (117528/124374)
94.4956341358
None
WORD-LEVEL:  Total accuracy: 79.188508 (15711/19840)
79.1885080645
None


### Character-based Bigram + Word-based Uni-gram model - KFTT-test

In [23]:
rawhyps = []
for id in xrange(1,1113): 
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper_lm_char_2).project(project_output=True)
    x2 = fst.compose(x1, cseq2wseq)
    ofst = fst.compose(x2, lm_word_1)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST))
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:

with open(KFTT_TESTCHARS) as f:
    kfttcrefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 95.220165 (123153/129335)
95.2201646886
None
WORD-LEVEL:  Total accuracy: 78.061593 (16197/20749)
78.0615933298
None


## Generate character-based bi-gram model lattice

### Character Lattice - PTB-dev

In [11]:
def num_arcs(f):
    return sum(f.num_arcs(s) for s in f.states())

In [12]:
prune_ptbdev_1 = fst.Fst.read('prune_ptbdev/1.fst')
print num_arcs(prune_ptbdev_1)

316


In [13]:
ptbdev_1 = fst.Fst.read('ptbdev.lats.lm.char.2/1.fst')
print num_arcs(ptbdev_1)

778


### Character Lattice - KFTT-dev

In [17]:
! mkdir -p prune.kfttdev.lats.lm.char.2/

for id in xrange(1,1114):
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst') # read input FSA with lower case character sequences
    # project to mixed case character sequences, apply the character bigram LM, and project onto the output
    x0 = fst.compose(ifst, toUpper_lm_char_2).project(project_output=True)
    x1 = fst.prune(x0, weight=10)
    fst.determinize(x1.rmepsilon()).minimize().arcsort().write('prune.kfttdev.lats.lm.char.2/'+str(id)+'.fst')

### Character Lattice - KFTT-test

In [24]:
! mkdir -p prune.kftttest.lats.lm.char.2/

for id in xrange(1,1113):
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper_lm_char_2).project(project_output=True)
    x1 = fst.prune(x0, weight=20)
    fst.determinize(x1.rmepsilon()).minimize().arcsort().write('prune.kftttest.lats.lm.char.2/'+str(id)+'.fst')

In [28]:
test = fst.Fst.read('kftttest.lats.lm.char.2/3.fst')
prune = fst.Fst.read('prune.kftttest.lats.lm.char.2/3.fst')
print num_arcs(test), num_arcs(prune)

72 264


### RNNLM + Char-Bi-gram model: PTB-dev

In [14]:
%%bash

source /home/ech57/tools/scripts/tf_helper_cpu.sh

python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1:100 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=101:200 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=201:300 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=301:400 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=401:500 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=501:600 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=601:700 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=701:800 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=801:900 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=901:1000 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1001:1100 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1101:1200 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1201:1300 \
        --output_path prune.ptbdev.lats.tolm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1301:1400 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1401:1500 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1501:1600 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1601:1700 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &


discarding /home/ech57/tools/anaconda/bin from PATH
prepending /home/ech57/tools/anaconda/envs/tensorflow_cpu/bin to PATH
2018-01-11 10:28:48,286 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,286 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,287 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,287 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,287 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,287 INFO: Added predictor fst with weight 1.0
2018-01-11 10:28:48,288 INFO: Settings from tensorflow config file:
2018-01-11 10:28:48,288 INFO: Settings from tensorflow config file:
2018-01-11 10:28:48,289 INFO: init_scale: 0.05
2018-01-11 10:28:48,289 INFO: learning_rate: 1.0
2018-01-11 10:28:48,289 INFO: Settings from tensorflow config file:
2018-01-11 10:28:48,289 INFO: max_grad_norm: 5
2018-01-11 10:28:48,289 INFO: num_layers: 2
2018-01-11 10:28:48,289 INFO: num_steps: 100
2018-01-11 10:28:48,289 INFO: init_scale: 0.05

In [22]:
rawhyps= []
for id in xrange(1,1701):
    ofst = fst.Fst.read('prune.ptbdev.lats.lm.char.2.rnnlm.sfst/'+str(id)+'.fst')
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST) )
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the PTB-dev character and word references:
with open(PTB_DEVCHARS) as f:
    ptbcrefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    ptbwrefs = [line.strip() for line in f]

print utilfst.evalrecasing(ptbcrefs, chyps)
print utilfst.evalrecasing(ptbwrefs, whyps)

Total accuracy: 99.343715 (207532/208903)
99.3437145469
None
Total accuracy: 97.617514 (33352/34166)
97.6175144881
None


In [21]:
ofst = fst.Fst.read('prune.ptbdev.lats.lm.char.2.rnnlm.sfst/1201.fst')
print num_arcs(ofst)

370


In [20]:
%%bash

source /home/ech57/tools/scripts/tf_helper_cpu.sh

python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1201:1220 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1221:1240 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1241:1260 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1261:1280 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1281:1300 \
        --output_path prune.ptbdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &

discarding /home/ech57/tools/anaconda/bin from PATH
prepending /home/ech57/tools/anaconda/envs/tensorflow_cpu/bin to PATH
2018-01-11 11:09:05,820 INFO: Added predictor fst with weight 1.0
2018-01-11 11:09:05,823 INFO: Settings from tensorflow config file:
2018-01-11 11:09:05,823 INFO: init_scale: 0.05
2018-01-11 11:09:05,823 INFO: learning_rate: 1.0
2018-01-11 11:09:05,823 INFO: max_grad_norm: 5
2018-01-11 11:09:05,823 INFO: num_layers: 2
2018-01-11 11:09:05,823 INFO: num_steps: 100
2018-01-11 11:09:05,824 INFO: hidden_size: 1000
2018-01-11 11:09:05,824 INFO: max_epoch: 1
2018-01-11 11:09:05,824 INFO: max_max_epoch: 15
2018-01-11 11:09:05,824 INFO: keep_prob: 0.5
2018-01-11 11:09:05,824 INFO: lr_decay: 0.8
2018-01-11 11:09:05,824 INFO: batch_size: 20
2018-01-11 11:09:05,824 INFO: vocab_size: 90
2018-01-11 11:09:05,828 INFO: Using LSTM cells of size=1000
2018-01-11 11:09:05,828 INFO: Model with 2 layer(s)
2018-01-11 11:09:05,829 INFO: Model with 1 unrolled step(s)
2018-01-11 11:09:05,88

### RNNLM + Char-Bi-gram model: KFTT-dev

In [23]:
%%bash

source /home/ech57/tools/scripts/tf_helper_cpu.sh

python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1:50 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=51:100 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=101:150 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=151:200 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=201:250 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=251:300 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=301:350 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=351:400 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=401:450 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=451:500 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=501:550 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=551:600 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=601:650 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=651:700 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=701:750 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=751:800 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=801:850 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=851:900 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=901:950 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=951:1000 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1001:1050 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1051:1113 \
        --output_path kfttdev.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &

discarding /home/ech57/tools/anaconda/bin from PATH
prepending /home/ech57/tools/anaconda/envs/tensorflow_cpu/bin to PATH
2018-01-11 11:19:07,774 INFO: Added predictor fst with weight 1.0
2018-01-11 11:19:07,775 INFO: Added predictor fst with weight 1.0
2018-01-11 11:19:07,777 INFO: Settings from tensorflow config file:
2018-01-11 11:19:07,777 INFO: Settings from tensorflow config file:
2018-01-11 11:19:07,777 INFO: init_scale: 0.05
2018-01-11 11:19:07,777 INFO: learning_rate: 1.0
2018-01-11 11:19:07,777 INFO: max_grad_norm: 5
2018-01-11 11:19:07,778 INFO: num_layers: 2
2018-01-11 11:19:07,777 INFO: Added predictor fst with weight 1.0
2018-01-11 11:19:07,778 INFO: num_steps: 100
2018-01-11 11:19:07,778 INFO: hidden_size: 1000
2018-01-11 11:19:07,778 INFO: max_epoch: 1
2018-01-11 11:19:07,778 INFO: max_max_epoch: 15
2018-01-11 11:19:07,778 INFO: keep_prob: 0.5
2018-01-11 11:19:07,778 INFO: lr_decay: 0.8
2018-01-11 11:19:07,778 INFO: batch_size: 20
2018-01-11 11:19:07,778 INFO: vocab_siz

In [42]:
rawhyps= []
for id in xrange(1,1114):
    ofst = fst.Fst.read('kfttdev.lats.lm.char.2.rnnlm.sfst/'+str(id)+'.fst')
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST) )
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the PTB-dev character and word references:
with open(KFTT_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]

print utilfst.evalrecasing(crefs, chyps)
print utilfst.evalrecasing(wrefs, whyps)

Total accuracy: 95.848811 (119211/124374)
95.8488108447
None
Total accuracy: 85.529234 (16969/19840)
85.529233871
None


### RNNLM + Char-Bi-gram model: KFTT-test

In [64]:
%%bash

source /home/ech57/tools/scripts/tf_helper_cpu.sh

python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1:50 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=51:100 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=101:150 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=151:200 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=201:250 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=251:300 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=301:350 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=351:400 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=401:450 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=451:500 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=501:550 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=551:600 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=601:650 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=651:700 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=701:750 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=751:800 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=801:850 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=851:900 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=901:950 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=951:1000 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1001:1050 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &
python /home/ech57/tools/sgnmt/decode.py \
        --config_file configs/sgnmt.rnnlm_chars.dev.ini --range=1051:1112 \
        --output_path kftttest.lats.lm.char.2.rnnlm.%s  --use_fst_weights true  --outputs=sfst &

discarding /home/ech57/tools/anaconda/bin from PATH
prepending /home/ech57/tools/anaconda/envs/tensorflow_cpu/bin to PATH
2018-01-10 21:05:32,876 INFO: Added predictor fst with weight 1.0
2018-01-10 21:05:32,879 INFO: Added predictor fst with weight 1.0
2018-01-10 21:05:32,879 INFO: Settings from tensorflow config file:
2018-01-10 21:05:32,879 INFO: Added predictor fst with weight 1.0
2018-01-10 21:05:32,880 INFO: init_scale: 0.05
2018-01-10 21:05:32,880 INFO: learning_rate: 1.0
2018-01-10 21:05:32,880 INFO: max_grad_norm: 5
2018-01-10 21:05:32,880 INFO: num_layers: 2
2018-01-10 21:05:32,880 INFO: num_steps: 100
2018-01-10 21:05:32,880 INFO: hidden_size: 1000
2018-01-10 21:05:32,880 INFO: max_epoch: 1
2018-01-10 21:05:32,880 INFO: max_max_epoch: 15
2018-01-10 21:05:32,880 INFO: keep_prob: 0.5
2018-01-10 21:05:32,880 INFO: lr_decay: 0.8
2018-01-10 21:05:32,880 INFO: batch_size: 20
2018-01-10 21:05:32,880 INFO: vocab_size: 90
2018-01-10 21:05:32,882 INFO: Settings from tensorflow config 

In [65]:
rawhyps= []
for id in xrange(1,1113):
    ofst = fst.Fst.read('kftttest.lats.lm.char.2.rnnlm.sfst/'+str(id)+'.fst')
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST) )
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the PTB-dev character and word references:
with open(KFTT_TESTCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    wrefs = [line.strip() for line in f]

print utilfst.evalrecasing(crefs, chyps)
print utilfst.evalrecasing(wrefs, whyps)

Total accuracy: 96.437159 (124727/129335)
96.437159315
None
Total accuracy: 84.375151 (17507/20749)
84.3751506097
None


# results


<table>
    <tr>
        <th>  </th>
        <th colspan="2">N-Gram</th>     
        <th colspan="2">TensorFlow</th>
        <th colspan="2">PTB-dev</th>
        <th colspan="2">KFTT-dev</th>
        <th colspan="2">KFTT-test</th>
    </tr>
    <tr>
        <th>  </th>
         <th>Char</th>
        <th>Word</th>
        <th colspan="2" ,align="center">RNNLM</th>
        <th colspan="2">Acc (%)</th>
        <th colspan="2">Acc (%)</th>
        <th colspan="2">Acc (%)</th>
    </tr>
    <tr>
        <th>ID</th>
        <th>Order</th>
        <th>Order</th>
         <th>Char</th>
        <th>Word</th>
        <th>Char</th>
        <th>Word</th>
        <th>Char</th>
        <th>Word</th>
        <th>Char</th>
        <th>Word</th>
    </tr>
    <tr>
          <th>1</th>
        <th>2</th>
        <th>-</th>
        <th>-</th>
        <th>-</th>
        <th>97.57</th>
        <th>87.56</th>
        <th>94.08</th>
        <th>76.44</th>
        <th> 94.81    </th> 
        <th> 75.48    </th> 
    </tr>
    <tr>
          <th>2</th>
        <th>-</th>
        <th>1</th>
        <th>-</th>
        <th>-</th>
        <th>96.61</th>
        <th>90.58</th>
        <th>?</th>
        <th>?</th>
        <th> ? </th> 
        <th> ?</th> 
    </tr>
     <tr>
          <th>3</th>
        <th>2</th>
        <th>1</th>
         <th>-</th>
        <th>-</th>
         <th>98.84</th>
        <th>94.80</th>
        <th>?</th>
        <th>?</th>
        <th> ? </th> 
        <th> ? </th> 
    </tr>
     <tr>
          <th>4</th>
        <th>2</th>
           <th>-</th>
        <th> X </th>
        <th>-</th>
        <th>99.39</th>
        <th>97.76</th>
        <th>?</th>
        <th>?</th>
        <th> ?    </th> 
        <th> ? </th> 
    </tr>
 
</table>

__Recasing Experiments Notes__
1. Character-level bigram language model (see Section 3).
2. Word-level unigram language model (see Section 4). 
3. Equal-weight interpolation of character-level bigram LMscores (from1.) with the word-level unigram LM scores (from 2.)
4. Character RNN LM rescoring of 2-gram character LM lattices.


## Points You May Wish to Consider in Your Report 

* How performance changes with character and word n-gram model order, and whether n-gram models can be made to work as well as character RNN LM.
* The speed of the various models and decoding strategies, and relationship to accuracy. 
* Accuracy and generalisation across the available data.
* Whether pruning and and weight pushing strategies can be used to speed up decoding. 
* Whether the performance metric is suitable for the task.


## Character-based N-gram model

In [69]:
! $DIR/tools/kenlm/build/bin/lmplz  -o 3 < $DIR/data/train/train.chars > lm.char.3.arpa

=== 1/5 Counting and sorting n-grams ===
Reading /remote/mlsalt-staff/wjb31/MLSALT/MLSALT3/practical/recasing/data/train/train.chars
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
****************************************************************************************************
Unigram tokens 4991911 types 89
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:1068 2:75235737600 3:141067010048
Statistics:
1 89 D1=0.5 D2=0.5 D3+=3
2 2889 D1=0.498538 D2=0.914263 D3+=1.49438
3 24930 D1=0.504988 D2=1.03759 D3+=1.61598
Memory estimate for binary LM:
type     kB
probing 508 assuming -p 1.5
probing 525 assuming -r models -p 1.5
trie    147 without quantization
trie     64 assuming -q 8 -b 8 quantization 
trie    146 assuming -a 22 array pointer compression
trie     62 assuming -a 22 -q 8 -b 8 array pointer compression and quantization
=== 3/5 Calculating and sorting initial probabilities ===
Chain sizes: 1:1068 2:46224 3:4

In [71]:
! LD_LIBRARY_PATH=$DIR/tools/kaldi/tools/openfst/lib/ \
  $DIR/tools/kaldi/src/lmbin/arpa2fst \
  --read-symbol-table=$DIR/data/chars.syms lm.char.3.arpa - > lm.char.3.fst

lm_char_3 = fst.Fst.read('lm.char.3.fst').arcsort()
toUpper_lm_char_3 = fst.compose(toUpper, fst.determinize(lm_char_3).minimize().arcsort())

/home/wjb31/MLSALT/MLSALT3/practical/recasing//tools/kaldi/src/lmbin/arpa2fst --read-symbol-table=/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/chars.syms lm.char.3.arpa - 
LOG (arpa2fst:Read():arpa-file-parser.cc:90) Reading \data\ section.
LOG (arpa2fst:HeaderAvailable():arpa-lm-compiler.cc:295) Reverting to slower state tracking because model is large: 3-gram with symbols up to 999999998
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \1-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \2-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \3-grams: section.


In [72]:
rawhyps = []
for id in xrange(1,1113): # KFTT-test has 1112 sentences
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_3)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_TESTCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 93.898790 (121444/129335)
93.898789964
None
WORD-LEVEL:  Total accuracy: 77.695311 (16121/20749)
77.6953106174
None


In [76]:
rawhyps = []
for id in xrange(1,1114): 
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_3)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 93.282358 (116019/124374)
93.2823580491
None
WORD-LEVEL:  Total accuracy: 78.805444 (15635/19840)
78.8054435484
None


In [77]:
rawhyps = []
for id in xrange(1,1701): 
    ifst = fst.Fst.read(PTB_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_3)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(PTB_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 98.081406 (204895/208903)
98.0814062029
None
WORD-LEVEL:  Total accuracy: 91.210560 (31163/34166)
91.2105602061
None


In [73]:
! $DIR/tools/kenlm/build/bin/lmplz  -o 4 < $DIR/data/train/train.chars > lm.char.4.arpa

=== 1/5 Counting and sorting n-grams ===
Reading /remote/mlsalt-staff/wjb31/MLSALT/MLSALT3/practical/recasing/data/train/train.chars
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
****************************************************************************************************
Unigram tokens 4991911 types 89
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:1068 2:36817485824 3:69032787968 4:110452465664
Statistics:
1 89 D1=0.5 D2=0.5 D3+=3
2 2889 D1=0.498538 D2=0.914263 D3+=1.49438
3 24930 D1=0.512808 D2=1.09784 D3+=1.73685
4 109218 D1=0.559102 D2=1.0145 D3+=1.52644
Memory estimate for binary LM:
type      kB
probing 2574 assuming -p 1.5
probing 2737 assuming -r models -p 1.5
trie     803 without quantization
trie     342 assuming -q 8 -b 8 quantization 
trie     777 assuming -a 22 array pointer compression
trie     316 assuming -a 22 -q 8 -b 8 array pointer compression and quantization
=== 3/5 Calculating and s

In [75]:
! LD_LIBRARY_PATH=$DIR/tools/kaldi/tools/openfst/lib/ \
  $DIR/tools/kaldi/src/lmbin/arpa2fst \
  --read-symbol-table=$DIR/data/chars.syms lm.char.4.arpa - > lm.char.4.fst

lm_char_4 = fst.Fst.read('lm.char.4.fst').arcsort()
toUpper_lm_char_4 = fst.compose(toUpper, fst.determinize(lm_char_4).minimize().arcsort())

rawhyps = []
for id in xrange(1,1113): # KFTT-test has 1112 sentences
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_4)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_TESTCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


/home/wjb31/MLSALT/MLSALT3/practical/recasing//tools/kaldi/src/lmbin/arpa2fst --read-symbol-table=/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/chars.syms lm.char.4.arpa - 
LOG (arpa2fst:Read():arpa-file-parser.cc:90) Reading \data\ section.
LOG (arpa2fst:HeaderAvailable():arpa-lm-compiler.cc:295) Reverting to slower state tracking because model is large: 4-gram with symbols up to 999999998
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \1-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \2-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \3-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \4-grams: section.
CHARACTER-LEVEL:  Total accuracy: 94.360382 (122041/129335)
94.3603819538
None
WORD-LEVEL:  Total accuracy: 80.456889 (16694/20749)
80.4568894887
None


In [78]:
rawhyps = []
for id in xrange(1,1114): 
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_4)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 94.077540 (117008/124374)
94.0775403219
None
WORD-LEVEL:  Total accuracy: 81.880040 (16245/19840)
81.8800403226
None


In [79]:
rawhyps = []
for id in xrange(1,1701): 
    ifst = fst.Fst.read(PTB_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_4)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(PTB_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 98.752531 (206297/208903)
98.752531079
None
WORD-LEVEL:  Total accuracy: 94.675994 (32347/34166)
94.6759936779
None


In [97]:
! $DIR/tools/kenlm/build/bin/lmplz  -o 6 < $DIR/data/train/train.chars > lm.char.6.arpa

=== 1/5 Counting and sorting n-grams ===
Reading /remote/mlsalt-staff/wjb31/MLSALT/MLSALT3/practical/recasing/data/train/train.chars
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
****************************************************************************************************
Unigram tokens 4991911 types 89
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:1068 2:13310938112 3:24958009344 4:39932817408 5:58235359232 6:79865634816
Statistics:
1 89 D1=0.5 D2=0.5 D3+=3
2 2889 D1=0.498538 D2=0.914263 D3+=1.49438
3 24930 D1=0.512808 D2=1.09784 D3+=1.73685
4 109218 D1=0.600289 D2=1.12659 D3+=1.56993
5 297081 D1=0.672407 D2=1.12406 D3+=1.52547
6 605766 D1=0.630934 D2=1.04957 D3+=1.43484
Memory estimate for binary LM:
type       kB
probing 20825 assuming -p 1.5
probing 23369 assuming -r models -p 1.5
trie     7557 without quantization
trie     3374 assuming -q 8 -b 8 quantization 
trie     6955 assuming -a 22 array poin

In [98]:
! LD_LIBRARY_PATH=$DIR/tools/kaldi/tools/openfst/lib/ \
  $DIR/tools/kaldi/src/lmbin/arpa2fst \
  --read-symbol-table=$DIR/data/chars.syms lm.char.6.arpa - > lm.char.6.fst

lm_char_6 = fst.Fst.read('lm.char.6.fst').arcsort()
toUpper_lm_char_6 = fst.compose(toUpper, fst.determinize(lm_char_6).minimize().arcsort())

rawhyps = []
for id in xrange(1,1113): # KFTT-test has 1112 sentences
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_6)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_TESTCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_TESTWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


/home/wjb31/MLSALT/MLSALT3/practical/recasing//tools/kaldi/src/lmbin/arpa2fst --read-symbol-table=/home/wjb31/MLSALT/MLSALT3/practical/recasing//data/chars.syms lm.char.6.arpa - 
LOG (arpa2fst:Read():arpa-file-parser.cc:90) Reading \data\ section.
LOG (arpa2fst:HeaderAvailable():arpa-lm-compiler.cc:295) Reverting to slower state tracking because model is large: 6-gram with symbols up to 999999998
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \1-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \2-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \3-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \4-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \5-grams: section.
LOG (arpa2fst:Read():arpa-file-parser.cc:145) Reading \6-grams: section.
CHARACTER-LEVEL:  Total accuracy: 95.537944 (123564/129335)
95.5379440987
None
WORD-LEVEL:  Total accuracy: 83.454624 (17316/20749)
83.4546243192
None


In [99]:
rawhyps = []
for id in xrange(1,1114): 
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_6)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(KFTT_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(KFTT_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 95.233731 (118446/124374)
95.2337305225
None
WORD-LEVEL:  Total accuracy: 84.581653 (16781/19840)
84.5816532258
None


In [100]:
rawhyps = []
for id in xrange(1,1701): 
    ifst = fst.Fst.read(PTB_DEVINPUT+str(id)+'.fst')
    ofst = fst.compose(ifst, toUpper_lm_char_6)
    rawhyps.append(utilfst.printstrings(ofst,syms=chars_ST,project_output=True)[0])
    
chyps = [l.replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(PTB_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy


CHARACTER-LEVEL:  Total accuracy: 99.277176 (207393/208903)
99.2771764886
None
WORD-LEVEL:  Total accuracy: 96.917989 (33113/34166)
96.9179886437
None


## Word-based N-Gram Language model

In [92]:
!sed -f $DIR/scripts/puncstrip.sed $DIR/data/train/train.words > lm.word.train.txt

! $DIR/tools/kenlm/build/bin/lmplz -o 4 < lm.word.train.txt > lm.word.4.arpa
    
! LD_LIBRARY_PATH=$DIR/tools/kaldi/tools/openfst/lib/ \
  $DIR/tools/kaldi/src/lmbin/arpa2fst \
  --read-symbol-table=$DIR/data/words.syms lm.word.4.arpa lm.word.4.fst    

lm_word_4 = fst.Fst.read('lm.word.4.fst')

=== 1/5 Counting and sorting n-grams ===
Reading /remote/mlsalt-2017/yw440/MLSALT3/practical/recasing/lm.word.train.txt
----5---10---15---20---25---30---35---40---45---50---55---60---65---70---75---80---85---90---95--100
****************************************************************************************************
Unigram tokens 815349 types 35027
=== 2/5 Calculating and sorting adjusted counts ===
Chain sizes: 1:420324 2:36817416192 3:69032656896 4:110452260864
Statistics:
1 35027 D1=0.59154 D2=1.09302 D3+=1.5078
2 331134 D1=0.783094 D2=1.11474 D3+=1.41668
3 614489 D1=0.897338 D2=1.26037 D3+=1.44564
4 700725 D1=0.940089 D2=1.37626 D3+=1.38432
Memory estimate for binary LM:
type       kB
probing 35369 assuming -p 1.5
probing 41047 assuming -r models -p 1.5
trie    16269 without quantization
trie     8881 assuming -q 8 -b 8 quantization 
trie    14813 assuming -a 22 array pointer compression
trie     7425 assuming -a 22 -q 8 -b 8 array pointer compression and quantization
=== 3/5 

In [ ]:
fst.prune(lm_word_4, weight=10)

In [94]:
rawhyps = []
for id in xrange(1,1114):
    ifst = fst.Fst.read(KFTT_DEVINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper).project(project_output=True) 
    x2 = fst.compose(x1, cseq2wseq) # compose with the chara83.01cter sequence to word sequence transducer
    ofst = fst.compose(x2, lm_word_4) # compose with the unigram language model
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST)) 

chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:
with open(KFTT_DEVCHARS) as f:
    kfttcrefs = [line.strip() for line in f]    
with open(KFTT_DEVWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL Total accuracy: 83.463586 (103807/124374)
83.4635856369
None
WORD-LEVEL Total accuracy: 76.018145 (15082/19840)
76.0181451613
None


In [93]:
rawhyps = []
for id in xrange(1,1701):
    ifst = fst.Fst.read(PTB_DEVINPUT+str(id)+'.fst')
    x1 = fst.compose(ifst, toUpper).project(project_output=True)
    x2 = fst.compose(x1, cseq2wseq)
    ofst = fst.compose(x2, lm_word_4)
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST) )
    
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-test character and word references:
with open(PTB_DEVCHARS) as f:
    crefs = [line.strip() for line in f]
with open(PTB_DEVWORDS) as f:
    wrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL: ', utilfst.evalrecasing(crefs, chyps) # character-level accuracy
print 'WORD-LEVEL: ', utilfst.evalrecasing(wrefs, whyps) # word-level accuracy

CHARACTER-LEVEL:  Total accuracy: 97.670211 (204036/208903)
97.6702105762
None
WORD-LEVEL:  Total accuracy: 96.397003 (32935/34166)
96.3970028683
None


In [95]:
rawhyps = []
for id in xrange(1,1113):
    ifst = fst.Fst.read(KFTT_TESTINPUT+str(id)+'.fst') # read input FSA with lower case character sequences
    x1 = fst.compose(ifst, toUpper).project(project_output=True) 
    x2 = fst.compose(x1, cseq2wseq) # compose with the character sequence to word sequence transducer
    ofst = fst.compose(x2, lm_word_4) # compose with the unigram language model
    rawhyps.append( utilfst.printstrings(ofst, syms=chars_ST)) 
chyps = [l[0].replace("<s>","").replace("</s>","") for l in rawhyps]
whyps = [line.replace(" ","").replace("_"," ") for line in chyps]

# Load the KFTT-dev character and word references:
with open(KFTT_TESTCHARS) as f:
    kfttcrefs = [line.strip() for line in f]    
with open(KFTT_TESTWORDS) as f:
    kfttwrefs = [line.strip() for line in f]
    
print 'CHARACTER-LEVEL', utilfst.evalrecasing(kfttcrefs, chyps) # character-level accuracy
print 'WORD-LEVEL', utilfst.evalrecasing(kfttwrefs, whyps) # word-level accuracy


CHARACTER-LEVEL Total accuracy: 82.318011 (106466/129335)
82.3180113658
None
WORD-LEVEL Total accuracy: 73.926454 (15339/20749)
73.926454287
None
